SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialSnapshotTest/SpatialSnapshotTest.cpp

```diff
@@ -0,0 +1,156 @@
+ *		- Stops map B.
+ *
+ * This means that we still are missing a crucial part of the test, we still didn't verify that the Snapshot loading works.
+ * Remember how (3) requires us to load a Snapshot with the map it was created, so we make the Automation Manager run
+ * these tests an even amount of times (more than 2 if you want to stress test it).
+ *
+ * The second time will be exact the same way as above, the 2 differences are that (a) map A will be launched with the
+ * Snapshot taken in the first run, and (b) ASpatialSnapshotTest will know that it is running from a custom Snapshot
+ * and will execute different steps.
+ */
+
+ASpatialSnapshotTest::ASpatialSnapshotTest()
+	: Super()
+{
+	Author = "Nuno";
+	Description = TEXT(
+		"Test SpatialOS Snapshots. This test is expected to run twice, the first time sets up the data and takes a Snapshot and the second "
+		"time loads from it and verifies the data is set.");
+	SetNumRequiredClients(1);
+}
+
+void ASpatialSnapshotTest::BeginPlay()
+{
+	Super::BeginPlay();
+
+	// First we need to know if we're launching from the default Snapshot or from a taken Snapshot.
+	bool bIsRunningFirstTime = !WasLoadedFromSnapshot();
+
+	FSpatialFunctionalTestStepDefinition VerifyDataStepDef = FSpatialFunctionalTestStepDefinition(true);
+	VerifyDataStepDef.StepName = TEXT("Verify Data Properly Set");
+	VerifyDataStepDef.TimeLimit = 5.0f;
+	VerifyDataStepDef.NativeTickEvent.BindLambda([this](float DeltaTime) {
+		ASpatialSnapshotTestActor* Actor = nullptr;
+		int NumActors = 0;
+		for (TActorIterator<ASpatialSnapshotTestActor> It(GetWorld()); It; ++It)
+		{
+			if (NumActors == 1)
+			{
+				FinishTest(EFunctionalTestResult::Failed, TEXT("There's more than one ASpatialSnapshotTestActor"));
+				return;
+			}
+			Actor = *It;
+			++NumActors;
+		}
+
+		if (IsValid(Actor))
+		{
+			// @TODO improve when we have The Verify functions
+			if (!Actor->VerifyBool())
+			{
+				return;
+			}
+			if (!Actor->VerifyInt32())
+			{
+				return;
+			}
+			if (!Actor->VerifyInt64())
+			{
+				return;
+			}
+			if (!Actor->VerifyFloat())
+			{
+				return;
+			}
+			if (!Actor->VerifyString())
+			{
+				return;
+			}
+			if (!Actor->VerifyName())
+			{
+				return;
+			}
+			if (!Actor->VerifyIntArray())
+			{
+				return;
+			}
+			FinishStep();
+		}
+	});
+
+	if (bIsRunningFirstTime)
+	{
+		// The first run we want to setup the data, verify it, and take snapshot.
+		AddStep(TEXT("First Run - Spawn Replicated Actor"), FWorkerDefinition::Server(1), nullptr, [this]() {
+			ASpatialSnapshotTestActor* Actor = GetWorld()->SpawnActor<ASpatialSnapshotTestActor>();
+
+			Actor->CrossServerSetProperties();
+
+			FinishStep();
+		});
+
+		FSpatialFunctionalTestStepDefinition FirstVerifyDataStepDef = VerifyDataStepDef;
+		FirstVerifyDataStepDef.StepName = FString::Printf(TEXT("%s - %s"), TEXT("First Run"), *VerifyDataStepDef.StepName);
+		AddStepFromDefinition(FirstVerifyDataStepDef, FWorkerDefinition::AllWorkers);
+
+		AddStep(
```

Comment:
This seems like it'll be pretty common functionality - can we simplify calling it?

Reply:
You mean taking snapshot? ATM we don't have any default step definitions in the SpatialFunctionalTest, but we can provide some. You think we should add a TakeSnapshotStepDefinition / ClearSnapshotStepDefinition?

Reply:
Great idea